src/GafferUSD/USDLayerWriter.cpp

```diff
@@ -68,6 +75,146 @@ using namespace GafferUSD;
+		deleteAttributes.addPaths( other.deleteAttributes, prefix );
+		deleteObject.addPaths( other.deleteObject, prefix );
+	}
+
+};
+
+// Recursively builds the PathMatchers held in a Filters struct, such that :
+//
+// - Objects with identical hashes will be included in `Filter::deleteObject`.
+// - Attributes with identical hashes will be included in `Filter::deleteAttributes`.
+// - Subtrees which are identical in all properties will be included in `Filter::prune`.
+Filters filtersWalk( const GafferScene::ScenePlug *baseScene, const GafferScene::ScenePlug *layerScene, const std::vector<float> &frames, const ScenePlug::ScenePath &path, const Gaffer::ThreadState &threadState, tbb::task_group_context &taskGroupContext )
+{
+	ScenePlug::PathScope pathScope( threadState, &path );
+
+	if( !layerScene->existsPlug()->getValue() )
+	{
+		return { false, g_emptyPathMatcher, g_emptyPathMatcher, g_emptyPathMatcher };
+	}
+
+	bool attributesMatch = true;
+	bool objectsMatch = true;
+	bool canPrune = true;
+
+	for( auto frame : frames )
+	{
+		pathScope.setFrame( frame );
+
+		if( attributesMatch )
+		{
+			attributesMatch = baseScene->attributesPlug()->hash() == layerScene->attributesPlug()->hash();
+			canPrune = canPrune && attributesMatch;
+		}
+
+		if( objectsMatch )
+		{
+			objectsMatch = baseScene->objectPlug()->hash() == layerScene->objectPlug()->hash();
+			canPrune = canPrune && objectsMatch;
+		}
+
+		if( canPrune )
+		{
+			canPrune = baseScene->transformPlug()->hash() == layerScene->transformPlug()->hash();
+		}
+
+		if( canPrune )
+		{
+			canPrune = baseScene->boundPlug()->hash() == layerScene->boundPlug()->hash();
+		}
+	}
+
+	ScenePlug::ScenePath prefix;
+	if( path.size() )
+	{
+		prefix.push_back( path.back() );
+	}
+
+	Filters result;
+	result.fullyPruned = canPrune;
+	if( attributesMatch )
+	{
+		result.deleteAttributes.addPath( prefix );
+	}
+	if( objectsMatch )
+	{
+		result.deleteObject.addPath( prefix );
+	}
+
+	IECore::ConstInternedStringVectorDataPtr baseChildNamesData = baseScene->childNamesPlug()->getValue();
+	const vector<InternedString> &baseChildNames = baseChildNamesData->readable();
+	if( !baseChildNames.empty() )
+	{
+		using ChildNameRange = tbb::blocked_range<std::vector<IECore::InternedString>::const_iterator>;
+		const ChildNameRange loopRange( baseChildNames.begin(), baseChildNames.end() );
+
+		Filters childFilters = tbb::parallel_reduce(
+
+			loopRange,
+
+			Filters(),
+
+			[&] ( const ChildNameRange &range, Filters x ) {
+
+				ScenePlug::ScenePath childPath = path;
+				childPath.push_back( InternedString() );
+				for( const auto &childName : range )
+				{
+					childPath.back() = childName;
+					const Filters childFilters = filtersWalk( baseScene, layerScene, frames, childPath, threadState, taskGroupContext );
+					x.add( childFilters );
+				}
+				return x;
+			},
+
+			[] ( Filters x, const Filters &y ) {
+
+				x.add( y );
+				return x;
+
+			},
+
+			taskGroupContext
+
+
+		);
```

Comment:
I think we've already talked about whether this could fit into a parallelGatherLocations style idiom, and you've concluded that not worrying about that is the best way to get this rolled out, which is totally reasonable. Might be worth a comment though - do you have any expectation about whether it's likely to improve performance to avoid the overhead of doing nested parallel_reduce? Or would it just be an aesthetic thing to make it fit in with other traversal?

Reply:
Added a comment in https://github.com/GafferHQ/gaffer/pull/6254/commits/f99b188a5018391ede14fb3d55867b8d4cfba63b. I don't think there would be any performance difference in USDLayerWriter either way - a `SceneAlgo` version would still use nested `parallel_reduce`, just as `parallelProcessLocations()` uses nested `parallel_for()`. But there could be benefits to using this approach to building PathMatchers elsewhere.